TestingLowerBounds/ForMathlib/LeftRightDeriv.lean

```diff
@@ -60,6 +60,65 @@ lemma leftDeriv_eq_rightDeriv (f : ℝ → ℝ) :
   ext x
   simp [leftDeriv_eq_rightDeriv_apply]
 
+@[simp]
+lemma rightDeriv_const (c : ℝ) : rightDeriv (fun _ ↦ c) = 0 := by
+  ext x
+  rw [rightDeriv_def, Pi.zero_apply]
+  exact derivWithin_const x _ c (uniqueDiffWithinAt_Ioi x)
+
+@[simp]
+lemma leftDeriv_const (c : ℝ) : leftDeriv (fun _ ↦ c) = 0 := by
+  simp_rw [leftDeriv_eq_rightDeriv, rightDeriv_const, Pi.zero_apply, neg_zero]
+  rfl
+
+lemma rightDeriv_linear (a : ℝ) : rightDeriv (fun x ↦ a * x) = fun _ ↦ a := by
```

Comment:
That should not be a lemma, I think?
You should add a simp lemma that states that `rightDeriv (fun x ↦ a * f x) = a * rightDeriv f`, and another one that proves `rightDeriv (fun x ↦ x) = fun _ ↦ 1`. Then simp should be able to prove this and `rightDeriv_linear` is not needed.

Reply:
```lean
@[simp]
lemma rightDeriv_const_mul (a : ℝ) {f : ℝ → ℝ} :
    rightDeriv (fun x ↦ a * f x) = fun x ↦ a * rightDeriv f x := by
  ext x
  by_cases ha : a = 0
  · simp [ha]
  by_cases hfx : DifferentiableWithinAt ℝ f (Ioi x) x
  · simp_rw [rightDeriv_def]
    rw [derivWithin_const_mul (uniqueDiffWithinAt_Ioi x)]
    exact hfx
  · rw [rightDeriv_of_not_differentiableWithinAt hfx, mul_zero,
      rightDeriv_of_not_differentiableWithinAt]
    refine fun h_diff ↦ hfx ?_
    have : f = fun x ↦ a⁻¹ * (a * f x) := by ext; simp [ha]
    rw [this]
    exact h_diff.const_mul _

@[simp]
lemma rightDeriv_id : rightDeriv id = fun _ ↦ 1 := by
  ext x
  rw [rightDeriv_def, derivWithin_id]
  exact uniqueDiffWithinAt_Ioi x

@[simp]
lemma rightDeriv_id' : rightDeriv (fun x ↦ x) = fun _ ↦ 1 := rightDeriv_id
```

Reply:
those might be useful as well:
```lean
@[simp]
lemma rightDeriv_neg {f : ℝ → ℝ} : rightDeriv (fun x ↦ - f x) = fun x ↦ - rightDeriv f x := by
  have : (fun x ↦ - f x) = fun x ↦ -1 * f x := by simp
  rw [this, rightDeriv_const_mul]
  simp

@[simp]
lemma leftDeriv_neg {f : ℝ → ℝ} : leftDeriv (fun x ↦ - f x) = fun x ↦ - leftDeriv f x := by
  simp [leftDeriv_eq_rightDeriv]

```